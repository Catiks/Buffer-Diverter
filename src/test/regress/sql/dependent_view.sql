drop schema if exists dependent_view cascade;
create schema dependent_view;
set current_schema = dependent_view;

-- test 1：视图
-- 创建基表
DROP TABLE IF EXISTS base_table;
CREATE TABLE base_table
(
  base_table_id INTEGER,
  base_table_field NUMERIC(10,4)
);
-- 插入数据
insert into base_table values (1,1);
insert into base_table values (2,2);

-- 创建第一层直接引用表字段的视图对象master_viewx
DROP VIEW IF EXISTS master_view1;
CREATE OR REPLACE VIEW master_view1 AS
  SELECT
    base_table_id AS id,
    base_table_field AS field
  FROM base_table;

DROP VIEW IF EXISTS master_view2;
CREATE OR REPLACE VIEW master_view2 AS
  SELECT
    base_table_field AS field
  FROM base_table;

DROP VIEW IF EXISTS master_view3;
CREATE OR REPLACE VIEW master_view3 AS
  SELECT
    base_table_id AS id
  FROM base_table;
-- 创建第二层间接引用表字段的视图对象dependent_viewx
DROP VIEW IF EXISTS dependent_view1;
CREATE OR REPLACE VIEW dependent_view1 AS
  SELECT
    id AS dependent_id,
    field AS dependent_field
  FROM master_view1;

DROP VIEW IF EXISTS dependent_view2;
CREATE OR REPLACE VIEW dependent_view2 AS
  SELECT
    id AS dependent_id
  FROM master_view1;

DROP VIEW IF EXISTS dependent_view3;
CREATE OR REPLACE VIEW dependent_view3 AS
  SELECT
    field AS dependent_field
  FROM master_view1;

DROP VIEW IF EXISTS dependent_view4;
CREATE OR REPLACE VIEW dependent_view4 AS
  SELECT
    field AS dependent_field1
  FROM master_view2;

DROP VIEW IF EXISTS dependent_view5;
CREATE OR REPLACE VIEW dependent_view5 AS
  SELECT
    id AS dependent_id
  FROM master_view3;
-- 创建第三层间接引用表字段的视图对象second_dependent_viewx
DROP VIEW IF EXISTS second_dependent_view1;
CREATE OR REPLACE VIEW second_dependent_view1 AS
  SELECT
    dependent_id AS sec_dependent_id,
    dependent_field AS sec_dependent_field
  FROM dependent_view1;

DROP VIEW IF EXISTS second_dependent_view2;
CREATE OR REPLACE VIEW second_dependent_view2 AS
  SELECT
    dependent_id AS sec_dependent_id
  FROM dependent_view2;

DROP VIEW IF EXISTS second_dependent_view3;
CREATE OR REPLACE VIEW second_dependent_view3 AS
  SELECT
    dependent_field AS sec_dependent_field
  FROM dependent_view3;

-- 初始valid字段为true
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;

-- 修改表字段类型，预期视图失效，查询无效视图成功
ALTER TABLE base_table ALTER COLUMN base_table_field TYPE VARCHAR(32);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from second_dependent_view3;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

-- 修改表字段长度，预期视图失效，查询无效视图成功
ALTER TABLE base_table ALTER COLUMN base_table_field TYPE VARCHAR(64);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from second_dependent_view3;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

-- 删除表字段，预期视图失效，查询无效视图报错
ALTER TABLE base_table DROP COLUMN base_table_field;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from second_dependent_view3;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

-- 重新插入表字段，预期失效的视图被查询时恢复有效，引用链上的父视图也恢复有效状态
ALTER TABLE base_table ADD base_table_field INTEGER;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from second_dependent_view3;
select * from dependent_view3;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

-- 删除整个表，预期报错
DROP TABLE base_table;
-- 删除整个表，预期成功
DROP TABLE base_table CASCADE;

-- test2：物化视图
DROP TABLE IF EXISTS base_table;
CREATE TABLE base_table
(
  base_table_id INTEGER,
  base_table_field NUMERIC(10,4)
);

-- 插入数据
insert into base_table values (1,1);
insert into base_table values (2,2);

DROP MATERIALIZED VIEW IF EXISTS master_view1;
CREATE MATERIALIZED VIEW master_view1 AS
  SELECT
    base_table_id AS id,
    base_table_field AS field
  FROM base_table;
  
-- 插入新数据
insert into base_table values (3,3);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;

-- 重命名表字段，预期修改后查询到原数据
ALTER TABLE base_table rename COLUMN base_table_id to id;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from master_view1;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

-- 修改表字段类型，预期修改后查询到原数据
ALTER TABLE base_table ALTER COLUMN id TYPE VARCHAR(32);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from master_view1;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

-- 删除表字段，预期修改后查询到原数据
ALTER TABLE base_table DROP COLUMN base_table_field;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from master_view1;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

-- 刷新物化视图，预期报错
REFRESH MATERIALIZED VIEW master_view1;

-- 重新添加表字段，刷新物化视图，预期查询到同步后数据
ALTER TABLE base_table ADD base_table_field INTEGER;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
REFRESH MATERIALIZED VIEW master_view1;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
select * from master_view1;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
\d base_table

--- clean
drop schema dependent_view cascade;
