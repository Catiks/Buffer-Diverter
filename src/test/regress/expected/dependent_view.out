drop schema if exists dependent_view cascade;
NOTICE:  schema "dependent_view" does not exist, skipping
create schema dependent_view;
set current_schema = dependent_view;
-- test 1：视图
-- 创建基表
DROP TABLE IF EXISTS base_table;
NOTICE:  table "base_table" does not exist, skipping
CREATE TABLE base_table
(
  base_table_id INTEGER,
  base_table_field NUMERIC(10,4)
);
-- 插入数据
insert into base_table values (1,1);
insert into base_table values (2,2);
-- 创建第一层直接引用表字段的视图对象master_viewx
DROP VIEW IF EXISTS master_view1;
NOTICE:  view "master_view1" does not exist, skipping
CREATE OR REPLACE VIEW master_view1 AS
  SELECT
    base_table_id AS id,
    base_table_field AS field
  FROM base_table;
DROP VIEW IF EXISTS master_view2;
NOTICE:  view "master_view2" does not exist, skipping
CREATE OR REPLACE VIEW master_view2 AS
  SELECT
    base_table_field AS field
  FROM base_table;
DROP VIEW IF EXISTS master_view3;
NOTICE:  view "master_view3" does not exist, skipping
CREATE OR REPLACE VIEW master_view3 AS
  SELECT
    base_table_id AS id
  FROM base_table;
-- 创建第二层间接引用表字段的视图对象dependent_viewx
DROP VIEW IF EXISTS dependent_view1;
NOTICE:  view "dependent_view1" does not exist, skipping
CREATE OR REPLACE VIEW dependent_view1 AS
  SELECT
    id AS dependent_id,
    field AS dependent_field
  FROM master_view1;
DROP VIEW IF EXISTS dependent_view2;
NOTICE:  view "dependent_view2" does not exist, skipping
CREATE OR REPLACE VIEW dependent_view2 AS
  SELECT
    id AS dependent_id
  FROM master_view1;
DROP VIEW IF EXISTS dependent_view3;
NOTICE:  view "dependent_view3" does not exist, skipping
CREATE OR REPLACE VIEW dependent_view3 AS
  SELECT
    field AS dependent_field
  FROM master_view1;
DROP VIEW IF EXISTS dependent_view4;
NOTICE:  view "dependent_view4" does not exist, skipping
CREATE OR REPLACE VIEW dependent_view4 AS
  SELECT
    field AS dependent_field1
  FROM master_view2;
DROP VIEW IF EXISTS dependent_view5;
NOTICE:  view "dependent_view5" does not exist, skipping
CREATE OR REPLACE VIEW dependent_view5 AS
  SELECT
    id AS dependent_id
  FROM master_view3;
-- 创建第三层间接引用表字段的视图对象second_dependent_viewx
DROP VIEW IF EXISTS second_dependent_view1;
NOTICE:  view "second_dependent_view1" does not exist, skipping
CREATE OR REPLACE VIEW second_dependent_view1 AS
  SELECT
    dependent_id AS sec_dependent_id,
    dependent_field AS sec_dependent_field
  FROM dependent_view1;
DROP VIEW IF EXISTS second_dependent_view2;
NOTICE:  view "second_dependent_view2" does not exist, skipping
CREATE OR REPLACE VIEW second_dependent_view2 AS
  SELECT
    dependent_id AS sec_dependent_id
  FROM dependent_view2;
DROP VIEW IF EXISTS second_dependent_view3;
NOTICE:  view "second_dependent_view3" does not exist, skipping
CREATE OR REPLACE VIEW second_dependent_view3 AS
  SELECT
    dependent_field AS sec_dependent_field
  FROM dependent_view3;
-- 初始valid字段为true
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | t
 master_view2           | v           | t
 master_view3           | v           | t
 dependent_view1        | v           | t
 dependent_view2        | v           | t
 dependent_view3        | v           | t
 dependent_view4        | v           | t
 dependent_view5        | v           | t
 second_dependent_view1 | v           | t
 second_dependent_view2 | v           | t
 second_dependent_view3 | v           | t
(12 rows)

-- 修改表字段类型，预期视图失效，查询无效视图成功
ALTER TABLE base_table ALTER COLUMN base_table_field TYPE VARCHAR(32);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | f
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | f
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | f
(12 rows)

select * from second_dependent_view3;
 sec_dependent_field 
---------------------
 1.0000
 2.0000
(2 rows)

select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | t
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | t
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | t
(12 rows)

\d base_table
          Table "dependent_view.base_table"
      Column      |         Type          | Modifiers 
------------------+-----------------------+-----------
 base_table_id    | integer               | 
 base_table_field | character varying(32) | 

-- 修改表字段长度，预期视图失效，查询无效视图成功
ALTER TABLE base_table ALTER COLUMN base_table_field TYPE VARCHAR(64);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | f
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | f
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | f
(12 rows)

select * from second_dependent_view3;
 sec_dependent_field 
---------------------
 1.0000
 2.0000
(2 rows)

select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | t
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | t
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | t
(12 rows)

\d base_table
          Table "dependent_view.base_table"
      Column      |         Type          | Modifiers 
------------------+-----------------------+-----------
 base_table_id    | integer               | 
 base_table_field | character varying(64) | 

-- 删除表字段，预期视图失效，查询无效视图报错
ALTER TABLE base_table DROP COLUMN base_table_field;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | f
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | f
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | f
(12 rows)

select * from second_dependent_view3;
ERROR:  The view second_dependent_view3 is invalid, please make it valid before operation.
HINT:  Please re-add missing table fields.
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | f
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | f
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | f
(12 rows)

\d base_table
  Table "dependent_view.base_table"
    Column     |  Type   | Modifiers 
---------------+---------+-----------
 base_table_id | integer | 

-- 重新插入表字段，预期失效的视图被查询时恢复有效，引用链上的父视图也恢复有效状态
ALTER TABLE base_table ADD base_table_field INTEGER;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | f
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | f
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | f
(12 rows)

select * from second_dependent_view3;
 sec_dependent_field 
---------------------
                    
                    
(2 rows)

select * from dependent_view3;
 dependent_field 
-----------------
                
                
(2 rows)

select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
        relname         | object_type | valid 
------------------------+-------------+-------
 base_table             | r           | t
 master_view1           | v           | t
 master_view2           | v           | f
 master_view3           | v           | t
 dependent_view1        | v           | f
 dependent_view2        | v           | f
 dependent_view3        | v           | t
 dependent_view4        | v           | f
 dependent_view5        | v           | t
 second_dependent_view1 | v           | f
 second_dependent_view2 | v           | f
 second_dependent_view3 | v           | t
(12 rows)

\d base_table
   Table "dependent_view.base_table"
      Column      |  Type   | Modifiers 
------------------+---------+-----------
 base_table_id    | integer | 
 base_table_field | integer | 

-- 删除整个表，预期报错
DROP TABLE base_table;
ERROR:  cannot drop table base_table because other objects depend on it
DETAIL:  view master_view2 depends on table base_table
view dependent_view4 depends on view master_view2
view master_view3 depends on table base_table
view dependent_view5 depends on view master_view3
view master_view1 depends on table base_table
view dependent_view3 depends on view master_view1
view second_dependent_view3 depends on view dependent_view3
view dependent_view1 depends on view master_view1
view second_dependent_view1 depends on view dependent_view1
view dependent_view2 depends on view master_view1
view second_dependent_view2 depends on view dependent_view2
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
-- 删除整个表，预期成功
DROP TABLE base_table CASCADE;
NOTICE:  drop cascades to 11 other objects
DETAIL:  drop cascades to view master_view2
drop cascades to view dependent_view4
drop cascades to view master_view3
drop cascades to view dependent_view5
drop cascades to view master_view1
drop cascades to view dependent_view3
drop cascades to view second_dependent_view3
drop cascades to view dependent_view1
drop cascades to view second_dependent_view1
drop cascades to view dependent_view2
drop cascades to view second_dependent_view2
-- test2：物化视图
DROP TABLE IF EXISTS base_table;
NOTICE:  table "base_table" does not exist, skipping
CREATE TABLE base_table
(
  base_table_id INTEGER,
  base_table_field NUMERIC(10,4)
);
-- 插入数据
insert into base_table values (1,1);
insert into base_table values (2,2);
DROP MATERIALIZED VIEW IF EXISTS master_view1;
NOTICE:  materialized view "master_view1" does not exist, skipping
CREATE MATERIALIZED VIEW master_view1 AS
  SELECT
    base_table_id AS id,
    base_table_field AS field
  FROM base_table;
  
-- 插入新数据
insert into base_table values (3,3);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | t
(2 rows)

-- 重命名表字段，预期修改后查询到原数据
ALTER TABLE base_table rename COLUMN base_table_id to id;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | t
(2 rows)

select * from master_view1;
 id | field  
----+--------
  1 | 1.0000
  2 | 2.0000
(2 rows)

select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | t
(2 rows)

\d base_table
      Table "dependent_view.base_table"
      Column      |     Type      | Modifiers 
------------------+---------------+-----------
 id               | integer       | 
 base_table_field | numeric(10,4) | 

-- 修改表字段类型，预期修改后查询到原数据
ALTER TABLE base_table ALTER COLUMN id TYPE VARCHAR(32);
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | f
(2 rows)

select * from master_view1;
 id | field  
----+--------
  1 | 1.0000
  2 | 2.0000
(2 rows)

select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | f
(2 rows)

\d base_table
          Table "dependent_view.base_table"
      Column      |         Type          | Modifiers 
------------------+-----------------------+-----------
 id               | character varying(32) | 
 base_table_field | numeric(10,4)         | 

-- 删除表字段，预期修改后查询到原数据
ALTER TABLE base_table DROP COLUMN base_table_field;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | f
(2 rows)

select * from master_view1;
 id | field  
----+--------
  1 | 1.0000
  2 | 2.0000
(2 rows)

select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | f
(2 rows)

\d base_table
     Table "dependent_view.base_table"
 Column |         Type          | Modifiers 
--------+-----------------------+-----------
 id     | character varying(32) | 

-- 刷新物化视图，预期报错
REFRESH MATERIALIZED VIEW master_view1;
ERROR:  The materialized view master_view1 is invalid, please make it valid before operation.
HINT:  Please re-add missing table fields.
-- 重新添加表字段，刷新物化视图，预期查询到同步后数据
ALTER TABLE base_table ADD base_table_field INTEGER;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | f
(2 rows)

REFRESH MATERIALIZED VIEW master_view1;
select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | t
(2 rows)

select * from master_view1;
 id | field 
----+-------
 1  |      
 2  |      
 3  |      
(3 rows)

select relname, object_type, valid from pg_object join pg_class on object_oid=oid and relnamespace = (select Oid from pg_namespace where nspname='dependent_view') order by object_oid;
   relname    | object_type | valid 
--------------+-------------+-------
 base_table   | r           | t
 master_view1 | m           | t
(2 rows)

\d base_table
          Table "dependent_view.base_table"
      Column      |         Type          | Modifiers 
------------------+-----------------------+-----------
 id               | character varying(32) | 
 base_table_field | integer               | 

--- clean
drop schema dependent_view cascade;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table base_table
drop cascades to materialized view master_view1
